sycl/include/CL/sycl/detail/kernel_desc.hpp

```diff
@@ -68,6 +67,8 @@ template <char...> struct KernelInfoData {
   static constexpr const char *getName() { return ""; }
 };
 
+using size_t = decltype((__builtin_strlen("")));
```

Comment:
Can we replace `size_t` with a fundamental type?

Reply:
We cannot really, The type of size_t is going to be dependent on the platform.

That said, a better definition is to use the actual definition of size-t, which is the result type of size_t.  So I'd suggest making this:

using size_t = decltype(sizeof(int));

Reply:
Done. Have put `using size_t = decltype(sizeof(int));`

Reply:
What I aiming at is to avoid adding type alias here. Can we use `int` type directly?

Reply:
> We cannot really, The type of size_t is going to be dependent on the platform.
> 
> That said, a better definition is to use the actual definition of size-t, which is the result type of size_t. So I'd suggest making this:
> 
> using size_t = decltype(sizeof(int));

Sorry, missed that comment. Do we really need `size_t` type in this header?

Reply:
Presumably we could just used 'unsigned long'  (or unsigned long long?) below, but it makes our implementation different than what gets standardized (though this is obviously in a different namespace).

Reply:
I think `size_t` isn't always an `int`.

Reply:
But in this header, we could just use 'unsigned long' as Erich suggests and not need size_t to be defined.

Reply:
> 
> 
> I think `size_t` isn't always an `int`.

In fact, it is NEVER `int`, since it needs to be unsigned! Depending on Linux/Windows and 64/32 bit, it can be: `unsigned int`, `unsigned long`, or `unsigned long long`.  By Standard, I think it could also be `unsigned short` and possibly even `unsigned char`.

Reply:
Sorry, I should have been more clear.
If I understand it correctly the only purpose of the type alias defined here is to be used here:
```c++
using size_t = decltype((__builtin_strlen("")));

// C++14 like index_sequence and make_index_sequence
// not needed C++14 members (value_type, size) not implemented
template <class T, T...> struct integer_sequence {};
template <size_t... I> using index_sequence = integer_sequence<size_t, I...>;
template <size_t N>
using make_index_sequence = __make_integer_seq<integer_sequence, size_t, N>;

template <typename T> struct KernelInfoImpl {
private:
  static constexpr auto n = __builtin_unique_stable_name(T);
  template <size_t... I>
  static KernelInfoData<n[I]...> impl(index_sequence<I...>) {
    return {};
  }
...
```

NOTE: to use this type alias users must use like `::cl::sycl::detail::size_t`, so it's not supposed to be used outside of SYCL headers.

Basically my question is: Do we have to use C+ `size_t` type or we can re-write the code to use `unsigned int` instead like this:

```c++
//using size_t = decltype((__builtin_strlen("")));

// C++14 like index_sequence and make_index_sequence
// not needed C++14 members (value_type, size) not implemented
template <class T, T...> struct integer_sequence {};
template <unsigned int... I> using index_sequence = integer_sequence<unsigned int, I...>;
template <unsigned int N>
using make_index_sequence = __make_integer_seq<integer_sequence, unsigned int, N>;

template <typename T> struct KernelInfoImpl {
private:
  static constexpr auto n = __builtin_unique_stable_name(T);
  template <unsigned int... I>
  static KernelInfoData<n[I]...> impl(index_sequence<I...>) {
    return {};
  }
...
```

Reply:
I'd suggest `unsigned long long` to get 64 bits on windows as well, but yes, that is perfectly apossible here.  Since this version of index_sequence/make_index_sequence seems to be sued just for the KernelInfoName it shouldn't matter.

Reply:
It looks like upper limit for the type is the length of the longest possible kernel name. Right?

Reply:
> It looks like upper limit for the type is the length of the longest possible kernel name. Right?

I think so.

Reply:
@bader, declaration of `size_t` type alias is eliminated. `unsigned long long` is used now.

Reply:
Thanks. Probably this is not that serious issues, I just thought that defining `size_t` type in `cl:sycl::detail` namespace might cause some unexpected conflicts.